numba/np/arrayobj.py

```diff
@@ -5439,3 +5439,92 @@ def impl(arr, axis1, axis2):
         return np.transpose(arr, axes_tuple)
 
     return impl
+
+
+@register_jitable
+def _take_along_axis_impl_set_ni(arr, indices, axis, Ni, Nk):
+    for i in range(len(Ni)):
+        Ni = tuple_setitem(Ni, i, arr.shape[i])
+    return _take_along_axis_impl(arr, indices, axis, Ni, Nk)
+
+
+@register_jitable
+def _take_along_axis_impl_set_nk(arr, indices, axis, Ni, Nk):
+    for i in range(len(Nk)):
+        Nk = tuple_setitem(Nk, i, arr.shape[axis + 1 + i])
+    return _take_along_axis_impl(arr, indices, axis, Ni, Nk)
+
+
+@register_jitable
+def _take_along_axis_impl_set_ni_nk(arr, indices, axis, Ni, Nk):
+    for i in range(len(Ni)):
+        Ni = tuple_setitem(Ni, i, arr.shape[i])
+    for i in range(len(Nk)):
+        Nk = tuple_setitem(Nk, i, arr.shape[axis + 1 + i])
+    return _take_along_axis_impl(arr, indices, axis, Ni, Nk)
+
+
+@register_jitable
+def _take_along_axis_impl(arr, indices, axis, Ni, Nk):
+    # Based on example code in
+    # https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html
+    J = indices.shape[axis]  # Need not equal M
+    out = np.empty(Ni + (J,) + Nk, arr.dtype)
+
+    for ii in np.ndindex(Ni):
+        for ii in np.ndindex(Ni):
+            for kk in np.ndindex(Nk):
+                np_s_ = (slice(None, None, None),)
+                a_1d = arr[ii + np_s_ + kk]
+                indices_1d = indices[ii + np_s_ + kk]
+                out_1d = out[ii + np_s_ + kk]
+                for j in range(J):
+                    out_1d[j] = a_1d[indices_1d[j]]
+    return out
+
+
+@overload(np.take_along_axis, jit_options={'boundscheck':True})
+def arr_take_along_axis(arr, indices, axis):
```

Comment:
This should check that `arr` and `indices` are array types, along with a unittest adding to check the type handling. I think `indices` also has to have an integer dtype as it's used as an index, appropriate type checks and unittests should be added for this.